channels_redis/core.py

```diff
@@ -316,12 +317,59 @@ def _setup_encryption(self, symmetric_encryption_keys):
                 x.decode("utf8") for x in
                 await connection.zrange(key, 0, -1)
             ]
-        # TODO: More efficient implementation (lua script per shard?)
+
+        connection_to_channels, channel_to_message = (
+            self._map_channel_to_connection(channel_names, message)
+        )
+
+        for connection_index, channel_names in connection_to_channels.items():
+            # Create a LUA script specific for this connection.
+            # Make sure to use the message specific to this channel, it is
+            # stored in channel_to_message dict and contains the
+            # __asgi_channel__ key.
+            group_send_lua = b""
+            for channel_name in channel_names:
+                channel_specific_message = channel_to_message[channel_name]
+
+                # Push the message to the required channel
+                group_send_lua += (
+                    b"redis.call('RPUSH', '%s', '%s')\n"
+                    % (channel_name.encode("utf8"), channel_specific_message)
+                )
+
+                # Set the message expiry
+                group_send_lua += (
+                    b"redis.call('EXPIRE', '%s', %d)\n"
+                    % (channel_name.encode("utf8"), self.expiry)
+                )
+
+            async with self.connection(connection_index) as connection:
+                await connection.eval(
+                    group_send_lua, keys=channel_names
+                )
```

Comment:
You're not using a pre-loaded Lua script here but rather just making one dynamically - have you considered using a pipeline instead? Have you done any basic speed tests between both?

Reply:
I have done some benchmarking against a single host against 3 group send methods
basic send, LUA and Pipeline in this gist https://gist.github.com/mmcardle/4e59943275cc1ad221b7bdade21cb45e

Reply:
Yup, I am happy that lua is a lot better!

Reply:
Thanks @andrewgodwin I am just trying to debug an issue with the LUA version of group_send, getting a lovely `Segmentation fault` when using it in my application, works fine with pipeline.
Seeing if i can reproduce it.

Reply:
Is there really no way to assign lua script content statically in redis and call it with arguments?

Reply:
Have debugged the `Segmentation fault` was to do with hiredis parsing the results and throwing a unicode decode error. Have updated the group_send to pass in the messages as arguments to the `connection.eval(...)` function and everything seems alot happier